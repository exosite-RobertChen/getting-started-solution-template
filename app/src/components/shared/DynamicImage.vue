<template>
  <img
    :alt="imageAlt"
    :src="imageSrc"
    :class="imageClass"
    :style="imageStyle"
    data-qa="DynamicImg"
    @error="handleError"
  />
</template>

<script>
export default {
  name: 'DynamicImage',
  props: {
    imageAlt: {
      type: String,
      required: true
    },
    imageSources: {
      type: Array,
      required: true
    },
    imageClass: {
      type: String,
      required: false,
      default: ''
    },
    imageStyle: {
      type: String,
      required: false,
      default: ''
    }
  },
  data() {
    return {
      imageIndex: 0
    }
  },
  computed: {
    imageSrc() {
      return this.imageSources[this.imageIndex]
    }
  },
  methods: {
    handleError() {
      if (this.imageIndex + 1 >= this.imageSources.length) return

      this.imageIndex += 1
    }
  }
}
</script>
