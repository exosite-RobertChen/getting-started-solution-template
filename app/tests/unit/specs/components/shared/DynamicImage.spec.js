import merge from 'lodash.merge'
import { shallowComponent } from '@/../tests/unit/common'
import DynamicImage from '@/components/shared/DynamicImage.vue'

const defaultProps = {
  propsData: {
    imageAlt: 'image alt',
    imageSources: ['source1.png', 'fallback.png']
  }
}

const factory = overrides =>
  shallowComponent(DynamicImage, merge(defaultProps, overrides))

describe('DynamicImage.vue', () => {
  let wrapper

  afterEach(() => wrapper.destroy())

  test('can be mounted', () => {
    wrapper = factory()
    expect(wrapper.contains(DynamicImage)).toBe(true)
  })

  test('matches snapshot', () => {
    wrapper = factory()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('when handleError method called, image index is changed', () => {
    wrapper = factory()
    expect(wrapper.vm.imageIndex).toBe(0)
    wrapper.vm.handleError()
    expect(wrapper.vm.imageIndex).toBe(1)
  })

  test('when image errors handleError method is called', () => {
    const handleErrorSpy = jest.fn()
    wrapper = factory({
      methods: {
        handleError: handleErrorSpy
      }
    })

    expect(handleErrorSpy).not.toHaveBeenCalled()
    wrapper.find('[data-qa=DynamicImg]').trigger('error')
    expect(handleErrorSpy).toHaveBeenCalledTimes(1)
  })
})
