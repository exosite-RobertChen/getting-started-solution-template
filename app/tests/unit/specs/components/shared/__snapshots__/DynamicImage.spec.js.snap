// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicImage.vue matches snapshot 1`] = `
<img
  alt="image alt"
  class=""
  data-qa="DynamicImg"
  src="source1.png"
/>
`;
